### PR TITLE
feat: introduced changelog modal on downloads

### PR DESCRIPTION
**Pull request by ovflowd**
## Description

This PR introduces the support of ChangelogModals within the Download's page of the redesigned Node.js Website

Note that this PR might increase the client bundle size 🤔 we should probably check how much of an increase it is.

## Validation

The "changelog button" should open a changelog with the correctly rendered MDX data.

### CODE REVIEW

**Comment by ovflowd** [comment]
Note that this version of the ChangelogModal does not include the avatars of all contributors that made a release happen. We still need to hook this into some internal utility for getting the names of all release authors.

I remember @richardlau or @targos mentioning this, but I might be remembering this incorrectly 🤔

File: pages/en/download/current.mdx [hunk 1]
```diff
@@ -13,6 +13,8 @@ subtitle: Download Node.js the way you want.
 <section>
 Node.js includes <LinkWithArrow href="https://npmjs.com">npm</LinkWithArrow> (<Release.NpmVersion />) and corepack.
 
+Read the changelog for <Release.ChangelogLink>this verion</Release.ChangelogLink>
```
Comment (mikeesto) [inline pages/en/download/current.mdx]:
```suggestion
Read the changelog for <Release.ChangelogLink>this version</Release.ChangelogLink>
```

File: pages/en/download/index.mdx [hunk 2]
```diff
@@ -13,6 +13,8 @@ subtitle: Download Node.js the way you want.
 <section>
 Node.js includes <LinkWithArrow href="https://npmjs.com">npm</LinkWithArrow> (<Release.NpmVersion />) and corepack.
 
+Read the changelog for <Release.ChangelogLink>this verion</Release.ChangelogLink>
```
Comment (mikeesto) [inline pages/en/download/index.mdx]:
```suggestion
Read the changelog for <Release.ChangelogLink>this version</Release.ChangelogLink>
```

File: pages/en/download/package-manager/current.mdx [hunk 3]
```diff
@@ -13,6 +13,8 @@ subtitle: Download Node.js the way you want.
 <section>
 Node.js includes <LinkWithArrow href="https://npmjs.com">npm</LinkWithArrow> (<Release.NpmVersion />) and corepack.
 
+Read the changelog for <Release.ChangelogLink>this verion</Release.ChangelogLink>
```
Comment (mikeesto) [inline pages/en/download/package-manager/current.mdx]:
```suggestion
Read the changelog for <Release.ChangelogLink>this version</Release.ChangelogLink>
```

File: pages/en/download/package-manager/index.mdx [hunk 4]
```diff
@@ -13,6 +13,8 @@ subtitle: Download Node.js the way you want.
 <section>
 Node.js includes <LinkWithArrow href="https://npmjs.com">npm</LinkWithArrow> (<Release.NpmVersion />) and corepack.
 
+Read the changelog for <Release.ChangelogLink>this verion</Release.ChangelogLink>
```
Comment (mikeesto) [inline pages/en/download/package-manager/index.mdx]:
```suggestion
Read the changelog for <Release.ChangelogLink>this version</Release.ChangelogLink>
```

File: pages/en/download/prebuilt-binaries/current.mdx [hunk 5]
```diff
@@ -13,6 +13,8 @@ subtitle: Download Node.js the way you want.
 <section>
 Node.js includes <LinkWithArrow href="https://npmjs.com">npm</LinkWithArrow> (<Release.NpmVersion />) and corepack.
 
+Read the changelog for <Release.ChangelogLink>this verion</Release.ChangelogLink>
```
Comment (mikeesto) [inline pages/en/download/prebuilt-binaries/current.mdx]:
```suggestion
Read the changelog for <Release.ChangelogLink>this version</Release.ChangelogLink>
```

File: pages/en/download/prebuilt-binaries/index.mdx [hunk 6]
```diff
@@ -13,6 +13,8 @@ subtitle: Download Node.js the way you want.
 <section>
 Node.js includes <LinkWithArrow href="https://npmjs.com">npm</LinkWithArrow> (<Release.NpmVersion />) and corepack.
 
+Read the changelog for <Release.ChangelogLink>this verion</Release.ChangelogLink>
```
Comment (mikeesto) [inline pages/en/download/prebuilt-binaries/index.mdx]:
```suggestion
Read the changelog for <Release.ChangelogLink>this version</Release.ChangelogLink>
```

File: pages/en/download/source-code/current.mdx [hunk 7]
```diff
@@ -7,12 +7,14 @@ subtitle: Download Node.js the way you want.
...
   I want the <Release.VersionDropdown /> version of the Node.js source code.
 
-  <Release.SourceButton kind="source" />
+  <Release.SourceButton />
 </section>
 
 <section>
 Node.js includes <LinkWithArrow href="https://npmjs.com">npm</LinkWithArrow> (<Release.NpmVersion />) and corepack.
 
+Read the changelog for <Release.ChangelogLink>this verion</Release.ChangelogLink>
```
Comment (mikeesto) [inline pages/en/download/source-code/current.mdx]:
```suggestion
Read the changelog for <Release.ChangelogLink>this version</Release.ChangelogLink>
```

File: pages/en/download/source-code/index.mdx [hunk 8]
```diff
@@ -7,12 +7,14 @@ subtitle: Download Node.js the way you want.
...
   I want the <Release.VersionDropdown /> version of the Node.js source code.
 
-  <Release.SourceButton kind="source" />
+  <Release.SourceButton />
 </section>
 
 <section>
 Node.js includes <LinkWithArrow href="https://npmjs.com">npm</LinkWithArrow> (<Release.NpmVersion />) and corepack.
 
+Read the changelog for <Release.ChangelogLink>this verion</Release.ChangelogLink>
```
Comment (mikeesto) [inline pages/en/download/source-code/index.mdx]:
```suggestion
Read the changelog for <Release.ChangelogLink>this version</Release.ChangelogLink>
```

File: util/getNodeJsChangelog.ts [hunk 9]
```diff
@@ -30,3 +30,22 @@ export const getNodejsChangelog = (version: string): string => {
...
+  const [, changelogAuthor] =
+    changelog.match(/^## .*? \([^)]+\)[,.] @(\S+)/m) || [];
+
+  return changelogAuthor || 'The Node.js Project';
+};
+
+export const getNodeJsChangelogSlug = (changelog: string) => {
+  // looking for the release header eg:
+  // ## 2016-03-08, Version 5.8.0 (Stable)
+  const [, changelogHeadinjg] = changelog.match(/^## (.*)$/m) || [];
```
Comment (mikeesto) [inline util/getNodeJsChangelog.ts]:
```suggestion
  const [, changelogHeading] = changelog.match(/^## (.*)$/m) || [];
```

File: util/getNodeJsChangelog.ts [hunk 10]
```diff
@@ -30,3 +30,22 @@ export const getNodejsChangelog = (version: string): string => {
...
+
+  return changelogHeadinjg || '';
```
Comment (mikeesto) [inline util/getNodeJsChangelog.ts]:
```suggestion
  return changelogHeading || '';
```

**Review by mikeesto — COMMENTED** [review]
On mobile there's a "X" in the top right to close the modal. I think it would be good to keep it for wider viewports too?

**Comment by canerakdas** [comment]
I compared the page source sizes for the two versions:

[With changelog](https://nodejs-lhta2c4o3-openjs.vercel.app/en/download) modal;
Source size: `193.85 KB`

[Beta](https://beta-node-js-org.vercel.app/en/download) without changelog modal;
Source size: `86.65 KB`

----

In addition, when I check with `@next/bundle-analyzer`, the gzipped dimensions are as follows;

Server `feat/changelog-modal`;

image

Server `main`;

image


Due to the nature of our page structure, it is not clear exactly what these parts affect. There seems to be an increase only on `../app/[locale]/[[...path]]/page.js`.

----

Client `feat/changelog-modal`;

image

Client `main`;

image

I'm not sure exactly how we can reduce the size here, but I'll check it out as well (if it's possible) 👀

**Comment by ovflowd** [comment]
> I compared the page source sizes for the two versions:
> 
> [With changelog](https://nodejs-lhta2c4o3-openjs.vercel.app/en/download) modal; Source size: `193.85 KB`
> 
> [Beta](https://beta-node-js-org.vercel.app/en/download) without changelog modal; Source size: `86.65 KB`
> 
> In addition, when I check with `@next/bundle-analyzer`, the gzipped dimensions are as follows;
> 
> Server `feat/changelog-modal`;
> 
> image
> Server `main`;
> 
> image
> Due to the nature of our page structure, it is not clear exactly what these parts affect. There seems to be an increase only on `../app/[locale]/[[...path]]/page.js`.
> 
> Client `feat/changelog-modal`;
> 
> image
> Client `main`;
> 
> image
> I'm not sure exactly how we can reduce the size here, but I'll check it out as well (if it's possible) 👀

If you can investigate that'd be great! It is indeed a big bundle increase, mostly (I assume) because it is also bundling the Shiki WASM and Languages and Theme on the client-side 🤔 

Can you check what's the bundle difference if we removed Shiki from the ChangelogModal?

**Comment by targos** [comment]
Something feels weird with how the page seems to disappear entirely with no UI element to close the modal.

**Comment by ovflowd** [comment]
> Something feels weird with how the page seems to disappear entirely with no UI element to close the modal.

Indeed, I assume we need to add an X button. Hmm. @canerakdas as I mentioned before, feel free to hijack this PR :)

**Comment by AugustinMauroy** [comment]
In relation to what @targos said, I think we could make the background around the modal more transparent to see website layout.

**Review by bmuenzenmeyer — CHANGES_REQUESTED** [review]
my opinion is that feature is not worth the code, bundle size, or UX experiences. why not link to the GitHub release?

it feels similar to #6369 - how will we resolve differences or deltas from the figma's vision and reality of implementation? do we still have the @nodejs/ux-and-design team? maybe we expand those ranks with some true UX professionals

**Comment by AugustinMauroy** [comment]
>  how will we resolve differences or deltas from the figma's vision and reality of implementation? 

I think we should open a discussion/issue to find a way forward with the Figma and potential new features that won't be on this design. And also to know how to adapt the design to the real condition.

**Comment by richardlau** [comment]
> my opinion is that feature is not worth the code, bundle size, or UX experiences. why not link to the GitHub release?

Stepping back a bit, what this is doing is trying to present the notes/lists of changes for the release to allow users of the site a way of finding out what changed or is notable about a particular release. In the redesign this is more or less covered by linking through to the blog post for the release. The blog posts are generated via [`scripts/release-post`](https://github.com/nodejs/nodejs.org/tree/main/scripts/release-post) based on content from the changelog over in the core nodejs/node repository.

If we look at the pre-redesign (i.e. current) website:
- https://nodejs.org/en/download neither links to the changelog nor the blog post.
- The download buttons on the homepage have links underneath to the changelogs on GitHub. These have been problematic because for some releases the file linked to is too large for GitHub to render in the web UI (see for example https://github.com/nodejs/nodejs.org/issues/4816 -- there was also a suggestion there to link to the GitHub release but that's also had size limitations).
- https://nodejs.org/en/about/previous-releases also links to the changelogs on GitHub.
- None of the pages AFAICT link to the blog posts.

So the redesign is already doing something the pre-redesign site did not -- i.e. it presents the changelog information by linking to the blog posts. I think we should keep linking to the blog posts in the redesign and do not need to repeat the information by linking to the changelogs.

**Comment by ovflowd** [comment]
Right, the intent of this PR was to solve the issue we have of changelog rendering on GitHub.

But it would only cover/handle cases for the current LTS and current releases, afaik.

Since these changelogs use the same content as from the blog posts, probably it is good enough to just have the link to the release post.

One of the other goals of this Modal, was to have a collection of GitHub avatars rendering, that showcase all the committers that made said release possible, and pretty much highlight their contributions.

I think we could still add this, but directly to the Download page, maybe below the header?

Wdyt @nodejs/nodejs-website 

Regarding lack of designers, well, at this point since we do not have anyone here volunteering that works in that field, we lack the experts to assess such cases.

(Also @richardlau I pinged you here because I recall you mentioning there was a way to get a list of all committers of a release, can you point me to the right direction?)

**Comment by richardlau** [comment]
> (Also @richardlau I pinged you here because I recall you mentioning there was a way to get a list of all committers of a release, can you point me to the right direction?)

👀 I don't recall that.

**Comment by AugustinMauroy** [comment]
> One of the other goals of this Modal, was to have a collection of GitHub avatars rendering, that showcase all the  committers that made said release possible, and pretty much highlight their contributions.
> 
> I think we could still add this, but directly to the Download page, maybe below the header?

@ovflowd I didn't really get your idea, you want to have an `AvatarGroup` on `BlogHeader` with list of contributor ? and set aside the modal?

**Comment by ovflowd** [comment]
> > One of the other goals of this Modal, was to have a collection of GitHub avatars rendering, that showcase all the  committers that made said release possible, and pretty much highlight their contributions.
> > I think we could still add this, but directly to the Download page, maybe below the header?
> 
> @ovflowd I didn't really get your idea, you want to have an `AvatarGroup` on `BlogHeader` with list of contributor ? and set aside the modal?

No. I want AvatarGroup on Download page, that's what I wrote on my comment

**Comment by ovflowd** [comment]
> > (Also @richardlau I pinged you here because I recall you mentioning there was a way to get a list of all committers of a release, can you point me to the right direction?)
> 
> 👀 I don't recall that.

Oh, maybe I talked about that with someone else, but I recall our generated CHANGELOG (GitHub) contains the name of the commuters from their GitHub accounts, do you know where we do get that? Maybe cc @nodejs/releasers know?

**Comment by richardlau** [comment]
> > > (Also @richardlau I pinged you here because I recall you mentioning there was a way to get a list of all committers of a release, can you point me to the right direction?)
> > 
> > 
> > 👀 I don't recall that.
> 
> Oh, maybe I talked about that with someone else, but I recall our generated CHANGELOG (GitHub) contains the name of the commuters from their GitHub accounts, do you know where we do get that? Maybe cc @nodejs/releasers know?

[changelog-maker](https://github.com/nodejs/changelog-maker) pulls the information out of the git commits.

**Comment by canerakdas** [comment]
While examining the size of the page here, I noticed that all releases and changelogs were added to the page because we added the changelogs to the release in reducer. To reduce the size of the page, I have removed the changelog from the release and made it accessible via `next-data` 🎉 

Source size:
`89.19 KB`

**Comment by ovflowd** [comment]
@canerakdas could you rebase? And @bmuenzenmeyer could you re-review?

**Comment by ovflowd** [comment]
@canerakdas I believe we need to reduce blur and add a close button. Also the changelog should only open after finishing fetching the data (IMO)

**Comment by canerakdas** [comment]
I think it looks better/accessible now, you can check it in the Vercel preview 👀 cc @ovflowd

Before:
image

After:
image

Before/Dark:
image

After/Dark:
image

**Comment by ovflowd** [comment]
Awesome! Would it also bee too much to ask you to check into what Richard mentioned ([changelog-maker](https://github.com/nodejs/changelog-maker) pulls the information out of the git commits.) for getting the full list of GitHub usernames for the "AvatarGroup" on the modal?

**Comment by ovflowd** [comment]
I believe we can merge this as it is; But we need reviews cc @bmuenzenmeyer @nodejs/nodejs-website

**Review by araujogui — APPROVED** [review]

**Review by bmuenzenmeyer — CHANGES_REQUESTED** [review]
small accessibility bug to fix


https://github.com/nodejs/nodejs.org/assets/298435/2ac68983-69d7-4c27-b96b-f09ce453fc06

**Comment by canerakdas** [comment]
Ah, I completely missed this, I solved it by wrapping it with a button. If we think we may encounter such situations more often, it seems like we can use [Radix primitive slot](https://www.radix-ui.com/primitives/docs/utilities/slot)

**Review by bmuenzenmeyer — APPROVED** [review]
🤝

**Pull request merged**